File: swift_mort/MORT/Models/AccountTrust.swift
import Foundation

enum AccountTrustLevel: Int, Codable, CaseIterable, Sendable {
    case basic = 0
    case secured = 1
    case affiliation = 2
    case digitalGovernmentCredential = 3
    case providerIdentity = 4
    case enhancedAdultScreening = 5

    var key: String { "TRUST_LEVEL_\(rawValue)" }

    var title: String {
        switch self {
        case .basic: "Basic account"
        case .secured: "Account secured"
        case .affiliation: "Affiliation verified"
        case .digitalGovernmentCredential: "Government digital ID verified"
        case .providerIdentity: "Provider identity verified"
        case .enhancedAdultScreening: "Enhanced adult screening"
        }
    }
}

struct TrustIndicator: Codable, Hashable, Identifiable, Sendable {
    let key: String
    let category: String
    let label: String
    let status: String
    let whatWasChecked: String
    let whatWasNotChecked: String
    let checkedAt: String?
    let expiresAt: String?
    let grantsMarketplaceAccess: Bool
    let doesNotGuaranteeSafety: Bool
    let environment: String?

    var id: String { "\(key)-\(environment ?? "current")" }

    enum CodingKeys: String, CodingKey {
        case key, category, label, status, environment
        case whatWasChecked = "what_was_checked"
        case whatWasNotChecked = "what_was_not_checked"
        case checkedAt = "checked_at"
        case expiresAt = "expires_at"
        case grantsMarketplaceAccess = "grants_marketplace_access"
        case doesNotGuaranteeSafety = "does_not_guarantee_safety"
    }
}

struct MarketplaceTrustEligibility: Codable, Hashable, Sendable {
    let allowed: Bool
    let action: String?
    let requiredLevel: Int
    let currentLevel: Int
    let missingRequirements: [String]
    let reasonCodes: [String]
    let retryAfter: String?
    let supportRoute: String
    let policyVersion: Int?
    let productionMarketplaceEnabled: Bool?
    let guardianModeOptional: Bool?
    let testMode: Bool?

    enum CodingKeys: String, CodingKey {
        case allowed, action
        case requiredLevel = "required_level"
        case currentLevel = "current_level"
        case missingRequirements = "missing_requirements"
        case reasonCodes = "reason_codes"
        case retryAfter = "retry_after"
        case supportRoute = "support_route"
        case policyVersion = "policy_version"
        case productionMarketplaceEnabled = "production_marketplace_enabled"
        case guardianModeOptional = "guardian_mode_optional"
        case testMode = "test_mode"
    }
}

struct AccountTrustContactStatus: Codable, Hashable, Sendable {
    let emailVerified: Bool
    let phoneVerified: Bool
    let phoneVerificationAvailable: Bool
    let emailOrPhoneIsLegalIdentity: Bool

    enum CodingKeys: String, CodingKey {
        case emailVerified = "email_verified"
        case phoneVerified = "phone_verified"
        case phoneVerificationAvailable = "phone_verification_available"
        case emailOrPhoneIsLegalIdentity = "email_or_phone_is_legal_identity"
    }
}

struct AccountSecuritySummary: Codable, Hashable, Sendable {
    let passkeyCount: Int
    let passkeysEnabledByServer: Bool
    let deviceBiometricsAreLocalAccountSecurityOnly: Bool
    let suspiciousSessionMonitoringEnabled: Bool

    enum CodingKeys: String, CodingKey {
        case passkeyCount = "passkey_count"
        case passkeysEnabledByServer = "passkeys_enabled_by_server"
        case deviceBiometricsAreLocalAccountSecurityOnly = "device_biometrics_are_local_account_security_only"
        case suspiciousSessionMonitoringEnabled = "suspicious_session_monitoring_enabled"
    }
}

struct AccountRiskSummary: Codable, Hashable, Sendable {
    let riskLevel: String
    let riskReasons: [String]
    let recommendedAction: String
    let humanReviewRequired: Bool
    let notACriminalAccusation: Bool

    enum CodingKeys: String, CodingKey {
        case riskLevel = "risk_level"
        case riskReasons = "risk_reasons"
        case recommendedAction = "recommended_action"
        case humanReviewRequired = "human_review_required"
        case notACriminalAccusation = "not_a_criminal_accusation"
    }
}

struct AccountTrustAvailability: Codable, Hashable, Sendable {
    let productionMarketplaceEnabled: Bool
    let identityDocumentCollectionEnabled: Bool
    let providerIdentityAvailable: Bool
    let appleWalletEnabled: Bool
    let androidDigitalCredentialsEnabled: Bool
    let manualAffiliationEvidenceCollectionEnabled: Bool

    enum CodingKeys: String, CodingKey {
        case productionMarketplaceEnabled = "production_marketplace_enabled"
        case identityDocumentCollectionEnabled = "identity_document_collection_enabled"
        case providerIdentityAvailable = "provider_identity_available"
        case appleWalletEnabled = "apple_wallet_enabled"
        case androidDigitalCredentialsEnabled = "android_digital_credentials_enabled"
        case manualAffiliationEvidenceCollectionEnabled = "manual_affiliation_evidence_collection_enabled"
    }
}

struct AccountTrustProfile: Codable, Hashable, Sendable {
    let ok: Bool
    let currentLevel: Int
    let levelKey: String
    let levelTitle: String
    let signalEnvironment: String
    let indicators: [TrustIndicator]
    let contactStatus: AccountTrustContactStatus
    let accountSecurity: AccountSecuritySummary
    let identityStatus: String
    let marketplaceEligibility: MarketplaceTrustEligibility
    let riskProfile: AccountRiskSummary
    let availability: AccountTrustAvailability
    let guardianModeOptional: Bool
    let schoolNamePublicByDefault: Bool
    let residentialAddressPublic: Bool
    let emailOrPhonePublic: Bool
    let peopleSearchUsed: Bool
    let safetyGuarantee: Bool
    let policyVersion: Int
    let calculatedAt: String?

    var level: AccountTrustLevel { AccountTrustLevel(rawValue: currentLevel) ?? .basic }

    enum CodingKeys: String, CodingKey {
        case ok, indicators
        case currentLevel = "current_level"
        case levelKey = "level_key"
        case levelTitle = "level_title"
        case signalEnvironment = "signal_environment"
        case contactStatus = "contact_status"
        case accountSecurity = "account_security"
        case identityStatus = "identity_status"
        case marketplaceEligibility = "marketplace_eligibility"
        case riskProfile = "risk_profile"
        case availability
        case guardianModeOptional = "guardian_mode_optional"
        case schoolNamePublicByDefault = "school_name_public_by_default"
        case residentialAddressPublic = "residential_address_public"
        case emailOrPhonePublic = "email_or_phone_public"
        case peopleSearchUsed = "people_search_used"
        case safetyGuarantee = "safety_guarantee"
        case policyVersion = "policy_version"
        case calculatedAt = "calculated_at"
    }
}

struct AccountTrustActionResult: Codable, Sendable {
    let ok: Bool
    let code: String?
    let message: String?
    let status: String?
    let requestID: UUID?
    let signalID: UUID?
    let membershipID: UUID?
    let checkID: UUID?
    let claimID: UUID?
    let appealID: UUID?
    let affiliationVerified: Bool?
    let identityVerified: Bool?
    let governmentIDVerified: Bool?
    let businessRecordMatched: Bool?
    let representativeIdentityVerified: Bool?
    let providerRequired: Bool?
    let grantsMarketplaceAccess: Bool?
    let publicVisibility: Bool?

    enum CodingKeys: String, CodingKey {
        case ok, code, message, status
        case requestID = "request_id"
        case signalID = "signal_id"
        case membershipID = "membership_id"
        case checkID = "check_id"
        case claimID = "claim_id"
        case appealID = "appeal_id"
        case affiliationVerified = "affiliation_verified"
        case identityVerified = "identity_verified"
        case governmentIDVerified = "government_id_verified"
        case businessRecordMatched = "business_record_matched"
        case representativeIdentityVerified = "representative_identity_verified"
        case providerRequired = "provider_required"
        case grantsMarketplaceAccess = "grants_marketplace_access"
        case publicVisibility = "public_visibility"
    }

    func requireSuccess(defaultMessage: String) throws -> AccountTrustActionResult {
        guard ok else { throw MortError.backend(code: code ?? "account_trust_action_failed", message: message ?? defaultMessage) }
        return self
    }
}

struct TrustAdminQueueResponse: Codable, Sendable {
    let ok: Bool
    let code: String?
    let queue: String?
    let items: [[String: JSONValue]]?
    let rawIdentityEvidenceIncluded: Bool?

    enum CodingKeys: String, CodingKey {
        case ok, code, queue, items
        case rawIdentityEvidenceIncluded = "raw_identity_evidence_included"
    }
}

File: flutter_mort/lib/data/models/account_trust.dart
enum AccountTrustLevel {
  basic(0, 'Basic account'),
  secured(1, 'Account secured'),
  affiliation(2, 'Affiliation verified'),
  digitalGovernmentCredential(3, 'Government digital ID verified'),
  providerIdentity(4, 'Provider identity verified'),
  enhancedAdultScreening(5, 'Enhanced adult screening');

  const AccountTrustLevel(this.value, this.title);

  factory AccountTrustLevel.fromValue(int value) {
    return values.firstWhere(
      (level) => level.value == value,
      orElse: () => AccountTrustLevel.basic,
    );
  }

  final int value;
  final String title;
}

class TrustIndicator {
  const TrustIndicator({
    required this.key,
    required this.category,
    required this.label,
    required this.status,
    required this.whatWasChecked,
    required this.whatWasNotChecked,
    required this.grantsMarketplaceAccess,
    required this.doesNotGuaranteeSafety,
    this.checkedAt,
    this.expiresAt,
    this.environment,
  });

  factory TrustIndicator.fromMap(Map<String, dynamic> map) {
    return TrustIndicator(
      key: _text(map['key'], fallback: 'unknown_signal'),
      category: _text(map['category'], fallback: 'other'),
      label: _text(map['label'], fallback: 'Trust signal'),
      status: _text(map['status'], fallback: 'unknown'),
      whatWasChecked: _text(
        map['what_was_checked'],
        fallback: 'No check description was returned.',
      ),
      whatWasNotChecked: _text(
        map['what_was_not_checked'],
        fallback: 'This signal does not guarantee identity or safety.',
      ),
      checkedAt: map['checked_at'] as String?,
      expiresAt: map['expires_at'] as String?,
      grantsMarketplaceAccess: map['grants_marketplace_access'] == true,
      doesNotGuaranteeSafety: map['does_not_guarantee_safety'] != false,
      environment: map['environment'] as String?,
    );
  }

  final String key;
  final String category;
  final String label;
  final String status;
  final String whatWasChecked;
  final String whatWasNotChecked;
  final String? checkedAt;
  final String? expiresAt;
  final bool grantsMarketplaceAccess;
  final bool doesNotGuaranteeSafety;
  final String? environment;
}

class MarketplaceTrustEligibility {
  const MarketplaceTrustEligibility({
    required this.allowed,
    required this.requiredLevel,
    required this.currentLevel,
    required this.missingRequirements,
    required this.reasonCodes,
    required this.supportRoute,
    required this.productionMarketplaceEnabled,
    required this.guardianModeOptional,
    required this.testMode,
    this.action,
    this.retryAfter,
    this.policyVersion,
  });

  factory MarketplaceTrustEligibility.fromMap(Map<String, dynamic> map) {
    return MarketplaceTrustEligibility(
      allowed: map['allowed'] == true,
      action: map['action'] as String?,
      requiredLevel: _integer(map['required_level']),
      currentLevel: _integer(map['current_level']),
      missingRequirements: _strings(map['missing_requirements']),
      reasonCodes: _strings(map['reason_codes']),
      retryAfter: map['retry_after'] as String?,
      supportRoute: _text(
        map['support_route'],
        fallback: '/support/account-trust',
      ),
      policyVersion: _nullableInteger(map['policy_version']),
      productionMarketplaceEnabled:
          map['production_marketplace_enabled'] == true,
      guardianModeOptional: map['guardian_mode_optional'] != false,
      testMode: map['test_mode'] == true,
    );
  }

  final bool allowed;
  final String? action;
  final int requiredLevel;
  final int currentLevel;
  final List<String> missingRequirements;
  final List<String> reasonCodes;
  final String? retryAfter;
  final String supportRoute;
  final int? policyVersion;
  final bool productionMarketplaceEnabled;
  final bool guardianModeOptional;
  final bool testMode;
}

class AccountTrustProfile {
  const AccountTrustProfile({
    required this.currentLevel,
    required this.levelKey,
    required this.levelTitle,
    required this.signalEnvironment,
    required this.indicators,
    required this.contactStatus,
    required this.accountSecurity,
    required this.identityStatus,
    required this.marketplaceEligibility,
    required this.riskProfile,
    required this.availability,
    required this.guardianModeOptional,
    required this.schoolNamePublicByDefault,
    required this.residentialAddressPublic,
    required this.emailOrPhonePublic,
    required this.peopleSearchUsed,
    required this.safetyGuarantee,
    required this.policyVersion,
    this.calculatedAt,
  });

  factory AccountTrustProfile.fromMap(Map<String, dynamic> map) {
    return AccountTrustProfile(
      currentLevel: _integer(map['current_level']),
      levelKey: _text(map['level_key'], fallback: 'TRUST_LEVEL_0'),
      levelTitle: _text(map['level_title'], fallback: 'Basic account'),
      signalEnvironment: _text(
        map['signal_environment'],
        fallback: 'production',
      ),
      indicators: _maps(map['indicators']).map(TrustIndicator.fromMap).toList(),
      contactStatus: _map(map['contact_status']),
      accountSecurity: _map(map['account_security']),
      identityStatus: _text(
        map['identity_status'],
        fallback: 'not_identity_verified',
      ),
      marketplaceEligibility: MarketplaceTrustEligibility.fromMap(
        _map(map['marketplace_eligibility']),
      ),
      riskProfile: _map(map['risk_profile']),
      availability: _map(map['availability']),
      guardianModeOptional: map['guardian_mode_optional'] != false,
      schoolNamePublicByDefault: map['school_name_public_by_default'] == true,
      residentialAddressPublic: map['residential_address_public'] == true,
      emailOrPhonePublic: map['email_or_phone_public'] == true,
      peopleSearchUsed: map['people_search_used'] == true,
      safetyGuarantee: map['safety_guarantee'] == true,
      policyVersion: _integer(map['policy_version']),
      calculatedAt: map['calculated_at'] as String?,
    );
  }

  final int currentLevel;
  final String levelKey;
  final String levelTitle;
  final String signalEnvironment;
  final List<TrustIndicator> indicators;
  final Map<String, dynamic> contactStatus;
  final Map<String, dynamic> accountSecurity;
  final String identityStatus;
  final MarketplaceTrustEligibility marketplaceEligibility;
  final Map<String, dynamic> riskProfile;
  final Map<String, dynamic> availability;
  final bool guardianModeOptional;
  final bool schoolNamePublicByDefault;
  final bool residentialAddressPublic;
  final bool emailOrPhonePublic;
  final bool peopleSearchUsed;
  final bool safetyGuarantee;
  final int policyVersion;
  final String? calculatedAt;

  AccountTrustLevel get level => AccountTrustLevel.fromValue(currentLevel);
  bool get passkeysEnabledByServer =>
      accountSecurity['passkeys_enabled_by_server'] == true;
  bool get productionMarketplaceEnabled =>
      availability['production_marketplace_enabled'] == true;
  bool get providerIdentityAvailable =>
      availability['provider_identity_available'] == true;
  bool get appleWalletEnabled => availability['apple_wallet_enabled'] == true;
  bool get androidDigitalCredentialsEnabled =>
      availability['android_digital_credentials_enabled'] == true;
}

Map<String, dynamic> _map(dynamic value) {
  return value is Map ? Map<String, dynamic>.from(value) : <String, dynamic>{};
}

List<Map<String, dynamic>> _maps(dynamic value) {
  return value is List
      ? value
            .whereType<Map>()
            .map((item) => Map<String, dynamic>.from(item))
            .toList()
      : <Map<String, dynamic>>[];
}

List<String> _strings(dynamic value) {
  return value is List ? value.whereType<String>().toList() : <String>[];
}

int _integer(dynamic value) => _nullableInteger(value) ?? 0;

int? _nullableInteger(dynamic value) => value is num ? value.toInt() : null;

String _text(dynamic value, {required String fallback}) {
  return value is String && value.trim().isNotEmpty ? value : fallback;
}
